음료 자판기 앱

GangWoon opened the pull request:
우유, 탄산, 커피 종류가 중첩되는 부분이 많아서 protocol을 채택할 건지 class를 상속받을 건지 고민했었습니다.

GangWoon committed:
프로젝트 생성

GangWoon committed:
ElementsOfBeverage 클래스 생성

다른 음료수 클래스들의 super class를 만들었습니다. protocol을 구현하려다가 jk가 말씀하신 "속성은 class로 method는 protocol로 "가 생각나서 바꿨습니다.

GangWoon committed:
BananaMilk 클래스 생성

ElementsOfBeverage를 상속받습니다.

GangWoon committed:
Cola 클래스 생성

ElementsOfBeverage를 상속받습니다.

GangWoon committed:
Top 클래스 생성

ElementsOfBeverage를 상속받습니다.

GangWoon committed:
ViewController 수정

각 Cola, Top, BananaMilk를 생성하고 출력했습니다.

GangWoon committed:
Info.plist 변경

Portrait를 사용하지 않도록 했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Models/ElementsOfBeverage.swift:
```diff
@@ -0,0 +1,33 @@
+    var weight: Int
+    var price: Int
+    var name: String
+    var date: Date
```
클래스 속성을 만들 때는 늘 private 을 우선적으로 고려하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Models/ElementsOfBeverage.swift:
```diff
@@ -0,0 +1,33 @@
+    var description: String {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyyMMdd"
+        return brand + ", " + String(weight) + "ml, " + String(price) + "원, " + name + ", " + dateFormatter.string(from: date)
```
- 날짜 형식을 바꾸는 것은 음료수의 기능이 아닌 것 같습니다. 다른 객체가 처리하도록 개선해보세요.
- 객체 속성을 문자열로 바꾸는 것이 단지 출력용으로 모든 단위, 표현식을 붙일 필요는 없습니다. 
모델 데이터를 형식에 맞춰 출력하는 것은 뷰에 따라 달라지고, 지역이나 언어에 따라도 달라집니다. 
그래서 이런 부분이 확장이 안되는 경직된 표현이 됩니다. 참고하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Models/ElementsOfBeverage.swift:
```diff
@@ -0,0 +1,33 @@
+        self.weight = weight
+        self.price = price
+        self.name = name
+        date = Date()
```
날짜는 무슨 날짜인가요? 좀 더 구체적으로 표현해보세요. 
그리고 항상 오늘날짜로 생성이 되어야 하나요? 초기값 하나도 유연하게 만드는 방법을 고민해보세요.

GangWoon replied:
명심하겠습니다

godrm commented on VendingMachineApp/VendingMachineApp/Models/Cola.swift:
```diff
@@ -0,0 +1,16 @@
+import Foundation
+
+class Cola: ElementsOfBeverage {
+    
```
빈 클래스는 선언하나 안하나 동일합니다. 
하위 클래스를 만든 다는 건 어떤 의미가 있을까요?
하위 클래스는 상위 클래스와 다르게 구체적인 속성이나 다른 메소드가 있어야 합니다. 그렇지 않으면 상위 클래스로 대체가 가능하니까요. 
하위 클래스에 들어갈 속성을 추가해보세요.

GangWoon replied:
추가했습니다~

godrm commented on VendingMachineApp/VendingMachineApp/Models/ElementsOfBeverage.swift:
```diff
@@ -0,0 +1,33 @@
+
+import Foundation
+
+class ElementsOfBeverage {
```
ElementsOfBeverage 는 어떤 의도가 있는건가요? 
굳이 번역해보자면 음료수 요소들? 음료수 원소들이 되나요?
어디에 속하는 원소들이 되는건가요? 선택했던 배경을 설명해주세요.

GangWoon replied:
음료수 요소들입니다! 변역기에서 찾았는데 중복되는 뜻이 있었군요!!

godrm replied:
번역기를 믿지마세요. 본인의 의도와 방향성을 담으세요.

godrm reviewed (commented):
본인은 신중하게 작성하는 것이겠지만, 제 지극히 개인적인 느낌으로 커밋이나 코드가 성의 없이 작성한 듯한 느낌이 있습니다. 
저도 어떤 부분이 그런 느낌을 주는지 명확하지 않습니다만 조심해야 할 부분인 것 같습니다. 
리뷰를 하는 이유 중에 하나는 개발 과정에서 고민과 함께 결과 코드를 함께 하는 동료와 코드를 읽고, 판단하고, 부품들을 조립해서 제품을 만들어가는 과정입니다. 
말했던 것처럼 저장소에 남은 코드들이 스스로 작업한 내용을 기록해놓는 겁니다. 
익숙한 것들이 잘못된 습관이거나 쉽게 넘어가는 것들이 오류가 있을 수도 있습니다. 
조금 더 세부적인 것까지 고민하고 판단하고 설계해서 적용해보면 좋겠습니다. 
그리고 그런 고민이 PR이나 커밋에도 반영되서 본인 저장소에 남기를 바랍니다. 
그런 맥락에서 피드백을 남겼습니다. 
이번 미션은 클래스 상속 관계를 연습하고, 어쩔수 없이 써야 하는 클래스 중심으로 설계하는 경험입니다. 
상위 클래스의 역할, 하위 클래스의 역할 - 그 구분까지 고민해보세요.

GangWoon committed:
DateGenerator 클래스 생성

DateGenerator라고 이름을 지은 이유는 TransitionDate와 고민하다가 날짜를 변경하는 거 같아서 DateGenerator가 조금 더 적당하다고 생각해서 지었습니다.
DateGenerator를 생성한 이유는 Date타입을 처리하기 위해서 만들었습니다.

GangWoon committed:
ElementsOfBeverage 클래스 수정

모든 속성을 private로 변경했습니다.
date타입을 처리하는 dateGenerator 객체를 생성했습니다. 더 이상 DateFormatter를 생성하지 않습니다.
additionalElements는 상품마다 추가적인 성분들을 표시할 속성입니다.
메소드 이름을 apply(featureOfBeverage:)로 지은 이유는 "additionalElements에 추가적인 내용을 적용한다. "라는 느낌으로 작성했습니다.
파라미터인 "featureOfBeverage"는 음료마다의 특징들을 표현했습니다.
각 음료마다 성분이 포함하고 있는 성분이 다르기 때문에 추가적으로 성분을 받을 수 있는 메소드를 만들었습니다.

GangWoon committed:
Top 클래스 수정

콜라의 특징인 carbonicAcid(탄산)이라는 속성을 추가했습니다.
apply메소드를 사용해서 추가적인 내용을 전달했습니다.

GangWoon committed:
Revert "Top 클래스 수정"

This reverts commit fd386ad954031da2c86fdd2b60076ce8ae5445cc.

GangWoon committed:
Top 클래스 수정

top의 특징인 coffeeVarieties를 속성으로 추가했습니다.
커피 품종이라는 의미로 사용했습니다.

GangWoon committed:
Cola 클래스 수정

콜라마다 pondusHydrogenii(PH)가 다를 수 있을거 같아서 속성으로 추가했습니다.

GangWoon committed:
BananaMilk 클래스 수정

다른 클래스에는 맛을 추가하지 않고 우유에다가 추가한 이유는 우유는 특정 맛으로 존재해서 추가했습니다.
flavor는 이름은 **맛으로 사용하기 위해서 지었습니다.

GangWoon committed:
ViewController 클래스 수정

cola, top, bananaMilk 클래스가 변경돼서 그에따른 수정을 했습니다.

GangWoon committed:
DateGenerator 파일이 생성하면서 새로 생긴 변경내용 같습니다.

GangWoon committed:
ViewController 클래스 수정

CustomStringConvertible에서 조금 더 유연하게 사용할 수 있도록 변경했습니다.

GangWoon commented:
수정했습니다~

godrm commented on VendingMachineApp/VendingMachineApp/Models/DateGenerator.swift:
```diff
@@ -0,0 +1,22 @@
+        let dateFormatter: DateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyyMMdd"
+        return dateFormatter
+    }()
```
새로운 클래스를 만들 때는 이미 있는 클래스를 확장하는 게 어떨까도 고민해보면 좋습니다.
Date와 가장 관련이 있고 응집도 높게 같이 있어야 한다면 Date를 확장하는 것도 좋겠죠?

GangWoon replied:
처음 작성할때는 date를 extension해서 작성했었지만 메소드 이름이 중복되는 느낌이라서 새로운 클래스를 만들었습니다. 생각하면서 작성하도록 노력하겠습니다!
